File: app/src/main/java/io/untaek/animal/UserDetailActivity.java
package io.untaek.animal;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Source;

import java.net.URI;
import java.util.ArrayList;
import java.util.List;

import io.untaek.animal.component.UserDetailListViewAdapter;
import io.untaek.animal.component.UserDetailListViewItem;
import io.untaek.animal.firebase.PostInTimeline;
import io.untaek.animal.firebase.UserDetail;
import io.untaek.animal.firebase.dummy;

public class UserDetailActivity extends AppCompatActivity {
    FirebaseFirestore db;
    DocumentReference docRef;


    List<List<PostInTimeline>> userPost = null;
    UserDetail user;
    String uploaderId = null;
    int userPosition = 0;

    List<UserDetailListViewItem> itemList;

    public UserDetailActivity() {
        db = FirebaseFirestore.getInstance();
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_detail);

        DocumentSnapshot documentSnapshot =(DocumentSnapshot)getIntent().getSerializableExtra("documentSnapshot");

        //uploaderId = getIntent().getExtras().getString("uploaderId");
        uploaderId = "dbsdlswp";

//        docRef = db.collection("users").document(uploaderId);
//        docRef.get().addOnCompleteListener(new OnCompleteListener<DocumentSnapshot>() {
//            @Override
//            public void onComplete(@NonNull Task<DocumentSnapshot> task) {
//                if(task.isSuccessful()){
//                    DocumentSnapshot document = task.getResult();
//                    textViewSetText(document);
//                    Log.d("ㅋㅋㅋ", "Cached document data : "+document.getData());
//
//                }else{
//                    Log.d("ㅋㅋㅋ","Cached get failed : "+task.getException());
//                }
//            }
//        });

        textViewSetText(documentSnapshot);

//        userPosition = 0;
//
//        for(UserDetail user : dummy.INSTANCE.getUsersDetail()){
//            if(user.getId().equals(uploaderId)){
//                break;
//            }
//            userPosition++;
//        }

        // itemlist 초기화 함수
//        defineItemList(dummy.INSTANCE.getUsersPost().get(userPosition));
//        user = dummy.INSTANCE.getUsersDetail().get(userPosition);


//        textViewSetText();
//        ListView listView = findViewById(R.id.user_detail_listView);
//
//        UserDetailListViewAdapter userDetailListViewAdapter = new UserDetailListViewAdapter(this, itemList);
//        listView.setAdapter(userDetailListViewAdapter);
    }

    List<UserDetailListViewItem> defineItemList(List<List<PostInTimeline>> user){
        itemList = new ArrayList<UserDetailListViewItem>();
        UserDetailListViewItem item;

        for(int i = 0; i<user.size(); i++){
            // petName을 이용하여 해당 pet의 게시글 가져오기
            item = new UserDetailListViewItem(user.get(i).get(0).getPetName(),user.get(i));
            itemList.add(item);
        }
        return  itemList;
    }

    private void textViewSetText(DocumentSnapshot documentSnapshot){
        TextView userName = findViewById(R.id.user_detail_user_name);
        ImageView userImage = findViewById(R.id.user_detail_user_image);
        TextView postCount = findViewById(R.id.user_detail_post_count);
        TextView followCount = findViewById(R.id.user_detail_follow_count);
        TextView likeCount = findViewById(R.id.user_detail_likes_count);


        userName.setText(documentSnapshot.get("name").toString());
        userImage.setImageURI(Uri.parse(documentSnapshot.get("photoURL").toString()));
        postCount.setText(documentSnapshot.get("total_posts").toString());
        followCount.setText(documentSnapshot.get("total_followers").toString());
        likeCount.setText(documentSnapshot.get("total_likes").toString());
    }

}

File: app/src/main/java/io/untaek/animal/UserDetailActivity.kt
package io.untaek.animal

import android.graphics.Color
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.OvalShape
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.util.Log
import android.widget.Toast
import com.bumptech.glide.Glide
import com.google.firebase.database.collection.LLRBNode
import io.untaek.animal.firebase.*
import io.untaek.animal.list.UserDetailRecyclerViewAdapter
import kotlinx.android.synthetic.main.activity_user_detail.*
import java.lang.Exception

class UserDetailActivity : AppCompatActivity(), Fire.Callback<UserDetail> {
    override fun onResult(data: UserDetail ) {
        user = UserDetail(user_.id, user_.name, user_.pictureUrl, data.totalLikes, data.totalPosts, data.totalFollowers)
        setWidget()
    }

    override fun onFail(e: Exception) {

        Log.e("ㅋㅋㅋ", "=================================onFail " + e)
    }

    private lateinit var user : UserDetail
    private lateinit var user_ : User
    private var followFlag   = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_detail)

        intent.getSerializableExtra("data")?.let { d ->
            user = d as UserDetail
        }
        intent.getSerializableExtra("postDetail_user")?.let { d ->
            user_ = d as User
            Fire.getInstance().getUserDetail(d.id, this)
            Log.e("ㅋㅋㅋ", "d name  : "+d.name)
            Log.e("ㅋㅋㅋ", "d picture  : "+d.pictureUrl)
            Log.e("ㅋㅋㅋ", "d id  : "+d.id)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        //viewer.release()
    }

    fun setWidget(){
        var userName = textView_userName_user_detail
        val userImage = imageView_userImage_user_detail
        val postCount  = textView_postsCount_user_detail
        val followerCount = textView_FollowersCount_user_detail
        val likeCount = textView_likesCount_user_detail

        Log.e("ㅋㅋㅋ", "user name : "+user.userName)
        Log.e("ㅋㅋㅋ", "user id  : "+user.id)
        Log.e("ㅋㅋㅋ", "user userimg : "+user.pictureUrl)
        Log.e("ㅋㅋㅋ", "user totalFollowers  : "+user.totalFollowers)
        Log.e("ㅋㅋㅋ", "user total post : "+user.totalPosts)
        Log.e("ㅋㅋㅋ", "user like  : "+user.totalLikes)
        userName.text = user.userName
        userImage.clipToOutline = true
        userImage.background = ShapeDrawable(OvalShape())

        Glide.with(this).load(user.pictureUrl).into(userImage)
        postCount.text = user.totalPosts.toString()
        likeCount.text = user.totalLikes.toString()
        followerCount.text = user.totalFollowers.toString()

        recyclerView_user_detail.layoutManager = GridLayoutManager(this, 3)
        recyclerView_user_detail.adapter = UserDetailRecyclerViewAdapter(this, user).also {
            it.update()
        }

        button_go_back_user_detail_detail.setOnClickListener { finish() }
        textView_followersLabel_user_detail.setOnClickListener {
            Log.e("ㅋㅋㅋ","userDetail Activity user.id : "+user.id)
            Fire.getInstance().checkFollow("dbsdlswp",user.id, object : Fire.Callback<Boolean> {
                override fun onResult(data: Boolean) {
                    if (data == true) {
                        Log.e("ㅋㅋㅋ", "unfollow")
                        Fire.getInstance().unfollow("dbsdlswp", user.id)
                        followerCount.text = (followerCount.text.toString().toInt() - 1).toString()
                    } else {
                        Log.e("ㅋㅋㅋ", "follow")
                        Fire.getInstance().follow("dbsdlswp", user.id)
                        followerCount.text = (followerCount.text.toString().toInt() + 1).toString()
                    }
                }

                override fun onFail(e: Exception) {
                    Log.e("ㅋㅋㅋ", "check follow 실패")
                }
            })
        }
    }
}
